[리액트 7기 김주빈] 라우팅 과제 제출

KJoobin opened the pull request:
필요할거같은 reducer, actions, api 먼저 구현을 했습니다!! 

구현할때는 괜찮은 방법이라고 생각했는데 PR 을 재출하려고 한번더 생각해보니깐 라우팅을 만들면서 필요할때 reducer, action 을 구현하는게 더 좋아보이는데 리듀서나 액션을 먼저 구현해도 괜찮을까요??

기존에 actions, reducer 가 구현되어있어서 크게 고민하지 않고 비슷하게 구현했던거 같습니다... 이부분은 반성해야할거같네요

KJoobin committed:
Install react-router-dom

KJoobin committed:
react-router-dom 초기 셋팅

KJoobin committed:
fetch restaurant 테스트

KJoobin committed:
fetch restaurant 구현

KJoobin committed:
reducer initial state 에 restaurant, selectedRestaurant 추가 테스트

KJoobin committed:
initial state 에 restaurant, selectedRestaurant 추가 구현

KJoobin committed:
setRestaurant reducer 테스트

KJoobin committed:
setRestaurant 구현

KJoobin committed:
select restaurant 테스트

KJoobin committed:
select restaurant 테스트 param 수정

KJoobin committed:
select resturant 테스트 initialState 수정

KJoobin committed:
select restaurant 구현

KJoobin committed:
install given2 & replace let to given

KJoobin committed:
fix typo

KJoobin committed:
select restuarnt 가 있을경우 테스트 + 구현

KJoobin committed:
load restaurant 테스트 typo 수정

KJoobin committed:
getState 호출 안하고 사용한거 수정

KJoobin committed:
console log 삭제

moonkii commented on src/actions.test.js:
```diff
@@ -36,48 +41,82 @@ describe('actions', () => {
   describe('loadRestaurants', () => {
     context('with selectedRegion and selectedCategory', () => {
       beforeEach(() => {
-        store = mockStore({
+        given('store', () => mockStore({
           selectedRegion: { id: 1, name: '서울' },
           selectedCategory: { id: 1, name: '한식' },
-        });
```
이렇게 각각 `beforeEach` 를 작성해 주면 given 이 굳이 필요하진 않을거에요.

moonkii commented on src/actions.test.js:
```diff
@@ -36,48 +41,82 @@ describe('actions', () => {
+        }));
+      });
+
+      it('run fetch restaurant and setRestaurant', async () => {
```
`runs` 가 맞겠죠?

moonkii reviewed (commented):
![image](https://user-images.githubusercontent.com/46776329/165121555-7a7f23cc-12d7-497e-b6c3-71017e09728d.png)


> 필요할거같은 reducer, actions, api 먼저 구현을 했습니다!!
구현할때는 괜찮은 방법이라고 생각했는데 PR 을 재출하려고 한번더 생각해보니깐 라우팅을 만들면서 필요할때 reducer, action 을 구현하는게 더 좋아보이는데 리듀서나 액션을 먼저 구현해도 괜찮을까요??

만약에 저라면 라우터로 기본적인 페이지를 구성해 놓은 뒤 한 페이지씩 완성해 나갈 것 같아요.
그렇지 않으면 페이지별로 확인하기도 어렵고, 테스트하기가 번거로울거에요.
강의에서 처럼 작은 단위로 하나씩 진행해 나가시면 좋아요.

> 기존에 actions, reducer 가 구현되어있어서 크게 고민하지 않고 비슷하게 구현했던거 같습니다... 이부분은 반성해야할거같네요

이미 비슷하게 만들어본 구조라서 테스트만 작성하고 그대로 구조를 만드셨을 것 같은데, 의미있는 HTML 마크업부터 시작해서 정말 하나씩 구조를 리펙터링해나면서 진행해보시면 TDD 를 연습하는데 더 도움이 되실거에요.

KJoobin committed:
HomePage About 링크 테스트

KJoobin committed:
render title 테스트

KJoobin committed:
render about 구현

KJoobin committed:
route about 테스트

KJoobin committed:
Restaurants link 테스트

KJoobin committed:
Restaurants link 구현

KJoobin committed:
About Page route 테스트, 구현

KJoobin committed:
render header 테스트

KJoobin committed:
헤더 구현

KJoobin committed:
헤더 클릭시 route / 테스트

KJoobin committed:
헤더 라우트 구현

KJoobin committed:
AboutPage text content 테스트

KJoobin committed:
route restaurants 페이지 테스트

KJoobin committed:
fix typo

KJoobin committed:
불필요한 beforeEach 삭제

KJoobin committed:
restaurants/[id] 라우트 테스트

KJoobin committed:
restaurant page route 에 추가

KJoobin committed:
restaurant page route 테스트

KJoobin committed:
restaurant page route 구현

KJoobin committed:
route with restaurant id 테스트

KJoobin committed:
route with params 구현

KJoobin committed:
render restaurant name 테스트

KJoobin committed:
redner restaurant name 구현

KJoobin committed:
render address test

KJoobin committed:
render 주소 구현

KJoobin committed:
메뉴 테스트

KJoobin committed:
메뉴 구현

KJoobin committed:
웹팩 devServer true 옵션 추가

KJoobin committed:
base url 추가

KJoobin commented:
fetch restaurant 설계가  잘못됨을 느껴서 리팩토링을 진행할 예정입니다..

KJoobin committed:
actions load restaurant 에서 store 삭제 테스트

KJoobin committed:
load restaurant params 추가

KJoobin committed:
Fix restaurant page 구현

moonkii commented on src/AboutPage.test.jsx:
```diff
@@ -0,0 +1,21 @@
+    expect(container).toHaveTextContent('About');
+  });
+
+  it('renders paragraph', () => {
```
paragraph 는 무슨 의미로 사용되나요?

KJoobin replied:
title, subtitle 과 같이 paragraph 라고 생각해서 썻는데, 햇깔릴수 있을거같아서 지웠습니다!

moonkii commented on src/App.test.jsx:
```diff
@@ -21,14 +20,71 @@ test('App', () => {
-  const { queryByText } = render((
-    <App />
+  const renderApp = () => render((
+    <MemoryRouter>
```
`initialEntries` 를 설정해주면 좋을 것 같아요.

moonkii commented on src/App.test.jsx:
```diff
@@ -21,14 +20,71 @@ test('App', () => {
+
+    expect(container).toHaveTextContent('헤더');
+  });
+
+  it('renders HomePage', () => {
+    const { container } = renderApp();
+
+    expect(container).toHaveTextContent('Home');
+    expect(container).toHaveTextContent('About');
+    expect(container).toHaveTextContent('Restaurants');
+  });
+
+  it('listens for click event on route /about', () => {
+    const { container, getByText } = renderApp();
+
+    fireEvent.click(getByText('About'));
+
+    expect(container).toHaveTextContent('About');
+    expect(container).toHaveTextContent('About 페이지 입니다.');
+  });
+
```
각각 route 마다 context 로 구분해주면 좋을 것 같아요.

KJoobin committed:
라우트 별로 context 로 분리

KJoobin committed:
AboutPage text 2개를 한개로 병합

KJoobin committed:
RestaurantPage 테스트 설명 수정

KJoobin committed:
Restaurant page todo 주석 추가

moonkii commented on src/RestaurantPage.test.jsx:
```diff
@@ -0,0 +1,60 @@
+    ));
+  });
+
+  context('when loading restaurant', () => {
```
여기서는 어떨때 loading 이되는지 맥락을 작성해주시면 좋을 것 같아요.

KJoobin replied:
음.. `페이지가 로드될때` 나, `페이지가 로드 되서 레스토랑 데이터를 불러오기 전` 과 같이 말씀하시는걸까요??

moonkii replied:
`given` 에서 드러나듯이 restaurant 가 없는 경우 로딩이 뜨는거죠.

KJoobin replied:
음!!  맞네요 레스토랑이 없을때 라고하는게 명확할거같아요!  

그런데 실질적으로는 레스토랑의 여부 보다는 /restaurants/1 과 같은 페이지에 접속했을때 로드를 하는건데 without restaurant 로 간략하게 써도 괜찮을까요?!

페이지에 접속 -> 레스토랑 데이터가 없음 -> 로딩 true + load 레스토랑데이터 -> 레스토랑 데이터를 불러오면 loading false

moonkii replied:
실제로도 without rastaurant 이니까 그렇게 작성하셔도 괜찮죠 ;)

KJoobin committed:
restaurant state 구조 변경 테스트

KJoobin committed:
restaurant initialState 구조 변경 구현

KJoobin committed:
set restaurant return value 구조 변경 테스트

KJoobin committed:
setRestaurant return value 구조 변경 구현

KJoobin committed:
NotFoundPage 테스트

KJoobin committed:
404 not found 구현

KJoobin committed:
without exist url 테스트

KJoobin committed:
without exist url 구현

KJoobin committed:
restaurant loading state 추가 테스트

KJoobin committed:
get Restaurant 테스트

KJoobin committed:
get restaurant reducer 구현

KJoobin committed:
getRestaurant 를 load restaurant 에 추가

KJoobin committed:
set restaurant initial state 수정 테스트

KJoobin committed:
RestaurantPage 에 loading, data 데이터 추가 테스트

KJoobin committed:
App test 에 restaurant initial state 수정

KJoobin committed:
RestaurantPage 에 restaurant state 구조 변경

- restaurant: restaurant -> restaurant: {data: restaurantData, loading: boolean, error:null}

KJoobin committed:
Restauran data 가 없을때 테스트

KJoobin committed:
restaurant data 가 없을때 구현

KJoobin committed:
restaurant id 가 없을경우 테스트

KJoobin committed:
restauarnt id 가 없을경우 구현

KJoobin committed:
RestaurantPage 를 RestaurantContainer 로 분리

KJoobin committed:
RestaurantDetail 테스트, 구현

KJoobin committed:
Restaurant Container 를 RestaurantDetail 과 분리

KJoobin committed:
Menu 가 없을경우 테스트, 구현

KJoobin committed:
메뉴가 있을경우 테스트, 구현

KJoobin committed:
Restaurant Detail 에 Menu Items 분리

moonkii commented:
해설강의를 참고하셔서 마무리해보시면 좋을 것 같아요 ;)
